azure/services/availabilitysets/availabilitysets.go

```diff
@@ -18,134 +18,108 @@ package availabilitysets
+	return err
 }
 
 // Delete deletes availability sets.
 func (s *Service) Delete(ctx context.Context) error {
 	ctx, log, done := tele.StartSpanWithLogger(ctx, "availabilitysets.Service.Delete")
 	defer done()
 
-	availabilitySetName, ok := s.Scope.AvailabilitySet()
-	if !ok {
-		return nil
-	}
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
 
-	as, err := s.Client.Get(ctx, s.Scope.ResourceGroup(), availabilitySetName)
-	if err != nil && azure.ResourceNotFound(err) {
-		// already deleted
+	setSpec := s.Scope.AvailabilitySetSpec()
+	if setSpec == nil {
+		log.V(2).Info("skip deletion when no availability set spec is found")
+		s.Scope.UpdateDeleteStatus(infrav1.AvailabilitySetReadyCondition, serviceName, nil)
 		return nil
 	}
 
+	existingSet, err := s.Client.Get(ctx, setSpec)
 	if err != nil {
-		return errors.Wrapf(err, "failed to get availability set %s in resource group %s", availabilitySetName, s.Scope.ResourceGroup())
+		if azure.ResourceNotFound(err) {
+			// already deleted
+			s.Scope.UpdateDeleteStatus(infrav1.AvailabilitySetReadyCondition, serviceName, nil)
+			return nil
+		}
+		resultingErr := errors.Wrapf(err, "failed to get availability set %s in resource group %s", setSpec.ResourceName(), setSpec.ResourceGroupName())
+
+		s.Scope.UpdateDeleteStatus(infrav1.AvailabilitySetReadyCondition, serviceName, resultingErr)
+		return resultingErr
 	}
+	availabilitySet, ok := existingSet.(compute.AvailabilitySet)
+	if !ok {
+		err = errors.Errorf("%T is not a compute.AvailabilitySet", existingSet)
 
-	// only delete when the availability set does not have any vms
-	if as.AvailabilitySetProperties != nil && as.VirtualMachines != nil && len(*as.VirtualMachines) > 0 {
-		return nil
+		s.Scope.UpdateDeleteStatus(infrav1.AvailabilitySetReadyCondition, serviceName, err)
+		return err
 	}
 
-	log.V(2).Info("deleting availability set", "availability set", availabilitySetName)
-	err = s.Client.Delete(ctx, s.Scope.ResourceGroup(), availabilitySetName)
-	if err != nil && azure.ResourceNotFound(err) {
-		// already deleted
+	// only delete when the availability set does not have any vms
+	if availabilitySet.AvailabilitySetProperties != nil && availabilitySet.VirtualMachines != nil && len(*availabilitySet.VirtualMachines) > 0 {
+		log.V(2).Info("skip deleting availability set with VMs", "availability set", setSpec.ResourceName())
+		s.Scope.UpdateDeleteStatus(infrav1.AvailabilitySetReadyCondition, serviceName, nil)
 		return nil
 	}
 
-	if err != nil {
-		return errors.Wrapf(err, "failed to delete availability set %s in resource group %s", availabilitySetName, s.Scope.ResourceGroup())
-	}
-
-	log.V(2).Info("successfully delete availability set", "availability set", availabilitySetName)
-
-	return nil
+	err = s.DeleteResource(ctx, setSpec, serviceName)
+	s.Scope.UpdateDeleteStatus(infrav1.AvailabilitySetReadyCondition, serviceName, err)
```

Comment:
it seems like there are 4 different places where we are calling `s.Scope.UpdateDeleteStatus`, can you find a way to reduce it to 1 or 2 times by calling it once when we return at the end, similar to what you did for Reconcile()?

Reply:
I just pushed some changes to only update the status at the end. Before, I structured it like that so it would be easy to read, just a linear series of `if` checks before deleting, and if any of them failed then we'd break out of it. I switched it to nested `if/else` blocks, do you think it's still intuitive to read?